feat(consensus): Misbehavior reports and verification

cyberphysic4l opened the pull request:
# Description of change

This PR deals with adding misbehavior reports for equivocation and provably faulty blocks to consensus blocks so we can get consensus on the associated misbehaviors. This PR does not deal with using the counts of these misbehaviors to take any actions, but they will be used to modify validator rewards in a future PR. 

## Equivocation reports
One type of misbehavior report implemented here is an equivocation report. Equivocation is the act of an authority signing and disseminating more than one block in a single consensus round which is a form of misbehavior. When an authority witnesses equivocation, they can report it by providing `BlockRef`s for two equivocating blocks in a `MisbehaviorReport` in the `misbehavior_reports` field of their next block. This misbehavior report can be thought of as similar to simple ancestors, except labelled to indicate misbehavior rather than providing approval for those blocks. 
## Provably faulty block reports
Another type of misbehavior report implemented here is `InvalidBlock` which reports a provably faulty block. A provably faulty block is one which passes signature verification, proving that it was created by that stated authority, but fails later verification check. Prior to this PR, such blocks were discarded upon receipt, but now they must be retained to be able to handle the reporting. As such, provably faulty blocks are now
stored in the DAG state. Provably faulty block reports consist simply of a `BlockRef` for the faulty block and are also included in the `misbehavior_reports` field of the block.

## 'Proving' misbehavior with misbehavior reports
Misbehavior reports do not really contain proof of the misbehavior itself, they are just a reference to the actual misbehavior and a statement that the issuer of the block containing the report claims to have witnessed the misbehavior. All references in the reports must be checked before ancestor selection, so it follows that anyone approving the block containing reports also claims to have witnessed those misbehaviors. To achieve this, when dealing with a block containing a misbehavior report, the `BlockRef`s in the report must be seen before accepting the block, and missing blocks from these refs must be retrieved before using that block for ancestor selection. This works in the same way as regular ancestors, ensuring that if a block containing an equivocation report is finalised, that means a quorum of validators witnessed the equivocation.

## Future use of misbehavior reports
Once the reports are present in the the DAG, they will be used to calculate a score for validators that will be used to adjust their rewards. The exact amount that the misbehaviors will affect rewards is yet to be determined.


## Links to any relevant issues

Be sure to reference any related issues by adding `fixes #(issue)`.

## How the change has been tested

- [ ] Basic tests (linting, compilation, formatting, unit/integration tests)
- [ ] Patch-specific tests (correctness, functionality coverage)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that new and existing unit tests pass locally with my changes

### Release Notes

- [x] Protocol: Adds misbehavior reporting to consensus blocks for provably faulty blocks and equivocations.
- [ ] Nodes (Validators and Full nodes):
- [ ] Indexer:
- [ ] JSON-RPC:
- [ ] GraphQL:
- [ ] CLI:
- [ ] Rust SDK:
- [ ] REST API:

cyberphysic4l committed:
Add equivocations to misbehavior reports

cyberphysic4l committed:
initial verification of misbehavior reports

cyberphysic4l committed:
verify misbehavior reports in block manager

cyberphysic4l committed:
add reporting of provably faulty blocks

cyberphysic4l committed:
persist faulty blocks in dag state storage

cyberphysic4l committed:
skip creation a VerifiedBlock for faulty blocks

cyberphysic4l committed:
Merge branch 'develop' into feat/misbehavior-reports

cyberphysic4l committed:
clippy lint

cyberphysic4l committed:
store provably faulty blocks

cyberphysic4l committed:
small fixes

cyberphysic4l committed:
fix: bug in get_blocks

cyberphysic4l committed:
Merge branch 'develop' into feat/misbehavior-reports

cyberphysic4l committed:
Add tests for block manager and dag state

cyberphysic4l committed:
clippy lints

cyberphysic4l committed:
Merge branch 'develop' into feat/misbehavior-reports

cyberphysic4l committed:
Merge branch 'develop' into feat/misbehavior-reports

cyberphysic4l committed:
Add metrics tracking

oliviasaa committed:
initial commit

cyberphysic4l committed:
clippy fix

polinikita commented on consensus/core/src/authority_service.rs:
```diff
@@ -395,8 +436,18 @@ impl<C: CoreThreadDispatcher> NetworkService for AuthorityService<C> {
             if !commit_sync_handle {
                 let all_ancestors = blocks
                     .iter()
+                    .filter_map(|b| {
+                        if let GetBlockResult::VerifiedBlock(b) = b {
+                            let mut ancestors = b.ancestors().to_vec();
+                            ancestors.extend(
+                                b.misbehavior_reports().iter().flat_map(|r| r.references()),
+                            );
+                            Some(ancestors)
+                        } else {
+                            None
+                        }
+                    })
```
We already don't go inside the if `if !self.context.protocol_config.consensus_batched_block_sync()` even on the Mainnet, so I would remove this part as the whole not if section should be deprecated.

cyberphysic4l replied:
I have to update this part because of the new `GetBlockResult` struct returned, so I think it's no harm to add the misbehavior reports there too. I have made sure to take care of it in the other branch too.

polinikita commented on consensus/core/src/block.rs:
```diff
@@ -620,15 +688,37 @@ impl TestBlock {
-    pub target: AuthorityIndex,
-    pub proof: MisbehaviorProof,
+    target: AuthorityIndex,
+    proof: MisbehaviorProof,
```
Not sure that this can be called a proof, but rather a hint as BlockRef is a digest+author+round.

cyberphysic4l replied:
Same comment as above, I agree about naming but would have to introduce a new block version in order to change this I think.

polinikita commented on consensus/core/src/block.rs:
```diff
@@ -620,15 +688,37 @@ impl TestBlock {
 pub struct MisbehaviorReport {
-    pub target: AuthorityIndex,
-    pub proof: MisbehaviorProof,
+    target: AuthorityIndex,
```
I am wondering whether target needs to be there. Would it be more natural to extract it from the BlockRef?

cyberphysic4l replied:
I would agree that it would be more natural to simply include a BlockRef, but my reason for using the existing struct is to avoid introducing a entirely new block version as `BlockV1` already includes a `Vec<MisbehaviorReport>` field.

polinikita commented:
Could you please write the scope of this PR? When arriving through the block stream. we are not adding blocks to the local DAG if they are provably faulty, but we do that if they arrive through the block synchronizer or commit syncer.  Also, what are the next planned PRs to make this misbehaviour mechanism fully working?

oliviasaa committed:
feat(iota-graphql-rpc, iota-json-rpc-types, iota-types, iota-execution): Introduction of ChangeEpochV4 transaction (#8522)

# Description of change

Introduces the ChangeEpochV4 type, which includes a vector of scores
relative to each authority behavior in the ending epoch.

## How the change has been tested

- [x] Basic tests (linting, compilation, formatting, unit/integration
tests)
- [ ] Patch-specific tests (correctness, functionality coverage)
- [ ] I have added tests that prove my fix is effective or that my
feature works
- [x] I have checked that new and existing unit tests pass locally with
my changes

### Release Notes

- [ ] Protocol:
- [x] Nodes (Validators and Full nodes): Introduces the ChangeEpochV4
type
- [ ] Indexer:
- [ ] JSON-RPC:
- [ ] GraphQL:
- [ ] CLI:
- [ ] Rust SDK:
- [ ] REST API:

oliviasaa committed:
feat(consensus, iota-core): Introduction of the Scorer component (#8530)

# Description of change

This PR is a refactoring of the metrics and scoring mechanism. All the
logic was moved from `metrics` to a `Scorer` component, owned by the
`Context` and the `AuthorityPerEpochStore` (wrapped inside an `Arc`):
- Consensus need access to the `Scorer` to update it accordingly to
blocks received and processed, so it's added to `Context`.
- Since the scores will be used in a new EndOfPublish message type, we
also add it to `AuthorityPerEpochStore`.


## How the change has been tested

- [x] Basic tests (linting, compilation, formatting, unit/integration
tests)
- [ ] Patch-specific tests (correctness, functionality coverage)
- [ ] I have added tests that prove my fix is effective or that my
feature works
- [x] I have checked that new and existing unit tests pass locally with
my changes

### Release Notes

- [ ] Protocol:
- [x] Nodes (Validators and Full nodes): Introduces the Scorer component
- [ ] Indexer:
- [ ] JSON-RPC:
- [ ] GraphQL:
- [ ] CLI:
- [ ] Rust SDK:
- [ ] REST API:

---------

Co-authored-by: Andrew Cullen <45826600+cyberphysic4l@users.noreply.github.com>

oliviasaa committed:
feat(iota-core, iota-types): Introduction of EndOfPublishV2 transaction  (#8529)

# Description of change

We introduce a new type of end of publish message (`EndOfPublishV2`)
containing a vector of `u64`. This vector represents the local view of
the authority about all validators' partial scores

## How the change has been tested

- [x] Basic tests (linting, compilation, formatting, unit/integration
tests)
- [ ] Patch-specific tests (correctness, functionality coverage)
- [ ] I have added tests that prove my fix is effective or that my
feature works
- [x] I have checked that new and existing unit tests pass locally with
my changes

### Release Notes

- [ ] Protocol:
- [x] Nodes (Validators and Full nodes): Introduces EndOfPublishV2
- [ ] Indexer:
- [ ] JSON-RPC:
- [ ] GraphQL:
- [ ] CLI:
- [ ] Rust SDK:
- [ ] REST API:

cyberphysic4l committed:
revert provable metrics changes

cyberphysic4l reviewed (commented):
> Could you please write the scope of this PR? 

I have improved the PR description to explain the scope of the PR. 

> When arriving through the block stream. we are not adding blocks to the local DAG if they are provably faulty, but we do that if they arrive through the block synchronizer or commit syncer.  

What do you mean by the block stream? Faulty blocks arriving via `handle_send_block` are added to the local DAG and used in they are added to the reports for the next block during smart ancestor selection. The faulty blocks also must be present in the DAG state in order to accept a block that reports them during the `handle_send_block`. 

> Also, what are the next planned PRs to make this misbehaviour mechanism fully working? 

This is now explained in the PR description.

cyberphysic4l committed:
retain faulty blocks in cash up until gc_round

polinikita commented on consensus/core/src/core.rs:
```diff
@@ -571,7 +586,7 @@ impl Core {
                 .protocol_config
                 .consensus_smart_ancestor_selection()
```
This is off for all networks. We decided to not use this feature and later on, it will not be available in Starfish. So, one needs to handle the else branch to include misbehaviour proofs

cyberphysic4l replied:
Thanks for spotting this, misbehavior reports are now dealt with in the regular ancestor selection aswell.

cyberphysic4l committed:
handle fetching provably faulty blocks in synchronizer

cyberphysic4l committed:
add misbehavior reports without smart ancestor selection

cyberphysic4l committed:
fmt

polinikita commented on consensus/core/src/synchronizer.rs:
```diff
@@ -625,28 +634,51 @@ impl<C: NetworkClient, V: BlockVerifier, D: CoreThreadDispatcher> Synchronizer<C
+        // Add any provably faulty blocks to core for processing
+        for provably_faulty_block in provably_faulty_blocks {
+            core_dispatcher
+                .add_provably_faulty_block(provably_faulty_block)
```
Could you please change the method to use as argument a vector of faulty blocks to align with add_blocks method.

cyberphysic4l replied:
Done

oliviasaa committed:
feat(consensus, iota-core, iota-protocol-config): Add scorer protocol config (#8618)

# Description of change

Adds the protocol configs to be used for the score integration to
consensus.

## How the change has been tested

- [x] Basic tests (linting, compilation, formatting, unit/integration
tests)
- [ ] Patch-specific tests (correctness, functionality coverage)
- [ ] I have added tests that prove my fix is effective or that my
feature works
- [x] I have checked that new and existing unit tests pass locally with
my changes

### Release Notes

- [ ] Protocol:
- [x] Nodes (Validators and Full nodes): New protocol configs relative
to scores
- [ ] Indexer:
- [ ] JSON-RPC:
- [ ] GraphQL:
- [ ] CLI:
- [ ] Rust SDK:
- [ ] REST API:

cyberphysic4l committed:
store faulty blocks on flush correctly

cyberphysic4l committed:
add vector of provably faulty blocks in core dispatcher

polinikita commented on consensus/core/src/block_manager.rs:
```diff
@@ -301,11 +311,13 @@ impl BlockManager {
-    fn verify_block_timestamps_and_accept(
+    fn verify_referenced_blocks_and_accept(
         &mut self,
         unsuspended_blocks: impl IntoIterator<Item = VerifiedBlock>,
```
could you please rename unsuspended_blocks to just accepted_or_unsuspended blocks as we extend accepted blocks with unsuspended at the previous step in the main flow

polinikita commented on consensus/core/src/block_manager.rs:
```diff
@@ -369,6 +384,32 @@ impl BlockManager {
+                    .verify_misbehavior_reports(&b, &proof_blocks)
+                {
+                    warn!("Block {:?} failed to verify misbehavior reports: {}", b, e);
+                    blocks_to_reject.insert(b.reference(), b);
```
This part seems to be problematic if the blocks are received through the commit syncer. We will not fetch these proof blocks through the synchroniser and the node will stuck forever. I suggest to add a flag `committed` from the previously called method and not check misbehaviour proofs in such a case.

cyberphysic4l committed:
Only check misbehavior reports if block is not committed

cyberphysic4l committed:
test misbehavior reports in committed blocks

cyberphysic4l committed:
Merge branch 'protocol-research/feat/score-integration' into feat/misbehavior-reports

cyberphysic4l committed:
store misbehavior reports in dag state and count in scorer

oliviasaa commented on consensus/core/src/storage/mem_store.rs:
```diff
@@ -208,6 +260,11 @@ impl Store for MemStore {
 
+    fn scan_misbehavior_reports(&self) -> ConsensusResult<HashSet<MisbehaviorReport>> {
+        let inner = self.inner.read();
+        Ok(inner.misbehavior_reports.clone())
```
checkk if this is not too much

oliviasaa commented on consensus/core/src/block.rs:
```diff
@@ -618,17 +686,39 @@ impl TestBlock {
+}
+
+impl MisbehaviorReport {
+    pub fn new(target: AuthorityIndex, proof: MisbehaviorProof) -> Self {
```
why passing the target?

oliviasaa commented on consensus/core/src/authority_service.rs:
```diff
@@ -108,21 +111,34 @@ impl<C: CoreThreadDispatcher> NetworkService for AuthorityService<C> {
+                    ])
+                    .inc();
+                info!("Unprovably faulty block from {}: {}", peer, e);
+                return Err(e);
```
why inside?

oliviasaa commented on consensus/core/src/authority_service.rs:
```diff
@@ -427,31 +456,43 @@ impl<C: CoreThreadDispatcher> NetworkService for AuthorityService<C> {
-                .flatten()
-                .collect()
+                .filter_map(|b| match b {
+                    GetBlockResult::VerifiedBlock(block) => Some(block.serialized().clone()),
```
why serialized?

oliviasaa commented on consensus/core/src/authority_service.rs:
```diff
@@ -473,31 +514,30 @@ impl<C: CoreThreadDispatcher> NetworkService for AuthorityService<C> {
-                if blocks.len() >= self.context.parameters.max_blocks_per_sync {
-                    blocks.truncate(self.context.parameters.max_blocks_per_sync);
+                let missing_blocks = dag_state
+                    .get_cached_blocks_in_range(
```
why not adding the faulty ones too?

cyberphysic4l replied:
Just because we don't store cached faulty blocks by authority and round as we do with verified blocks, so it is not possible at present to retrieve the missing ones in this way.

oliviasaa commented on consensus/core/src/block_manager.rs:
```diff
@@ -369,9 +385,38 @@ impl BlockManager {
+                            }
+                            MisbehaviorProof::Equivocation { .. } => {
+                                let report_proof_blocks =
+                                    (self.dag_state.read().get_blocks(&report.references()))
```
what if you don't find it?

oliviasaa commented:
I deleted my old comments because I pushed them by mistake, they were just local notes to myself.
I'm going to start now by adding short comments about small parts of the code

oliviasaa commented on consensus/core/src/storage/mem_store.rs:
```diff
@@ -208,6 +260,11 @@ impl Store for MemStore {
 
+    fn scan_misbehavior_reports(&self) -> ConsensusResult<HashSet<MisbehaviorReport>> {
+        let inner = self.inner.read();
+        Ok(inner.misbehavior_reports.clone())
```
During recovery, not everything is scanned, in general. I am not sure if reading all `MisbehaviorReport` from storage is doable. I did something similar, initially, for the unprovable metrics, and got reviews saying that this should not be done.

oliviasaa replied:
Same comment in rocksdb_store.rs

oliviasaa commented on consensus/core/src/authority_service.rs:
```diff
@@ -395,8 +411,18 @@ impl<C: CoreThreadDispatcher> NetworkService for AuthorityService<C> {
                 let all_ancestors = blocks
                     .iter()
+                    .filter_map(|b| {
+                        if let GetBlockResult::VerifiedBlock(b) = b {
```
another minor comment: `GetBlockResult` sounds odd to me, but I still don't have a better suggestion. I'll get back to this when I think of something

cyberphysic4l replied:
How about `BlockRetrievalResult`?

oliviasaa commented on consensus/core/src/authority_service.rs:
```diff
@@ -427,31 +456,43 @@ impl<C: CoreThreadDispatcher> NetworkService for AuthorityService<C> {
+                .iter()
+                .filter_map(|b| match b {
+                    GetBlockResult::VerifiedBlock(block) => Some(block.serialized().clone()),
+                    GetBlockResult::ProvablyFaultyBlock(block) => Some(block.serialized.clone()),
```
This should be checked with consensus, but, as I mentioned to you during a call, I wonder how much consensus would be impacted if we don't look for `ProvablyFaultyBlock` in storage for requests that came from the synchronizer (and only return blocks from cache). This would probably not work with the current logic and would require some changes, I think

oliviasaa commented on consensus/core/src/authority_service.rs:
```diff
@@ -473,31 +514,30 @@ impl<C: CoreThreadDispatcher> NetworkService for AuthorityService<C> {
+                        self.context
+                            .parameters
+                            .max_blocks_per_sync
+                            .saturating_sub(verified_blocks.len()),
```
I think here maybe it should be `bytes.len()`, no?

cyberphysic4l committed:
use sorted vector for equivocation report refs

cyberphysic4l committed:
save block refs only for misbehavior reports in dag state

oliviasaa commented on consensus/core/src/block.rs:
```diff
@@ -618,17 +686,50 @@ impl TestBlock {
+    pub fn new_invalid_block_report(block_ref: BlockRef) -> Self {
+        let target = block_ref.author;
+        let proof = MisbehaviorProof::InvalidBlock(block_ref);
+        Self { target, proof }
```
You have the target inside the `MisbehaviorProof`, no? Why passing it then?
Also, I might have missed where you do this, but I think the `target` in `MisbehaviorReport` is never verified agains the one in `MisbehaviorProof`

oliviasaa commented on consensus/core/src/block_verifier.rs:
```diff
@@ -237,6 +247,68 @@ impl BlockVerifier for SignedBlockVerifier {
+            match misbehavior_report.proof() {
+                MisbehaviorProof::InvalidBlock(_block_ref) => {
+                    // We just need to check that a provably faulty block was provided.
+                    if let Some(GetBlockResult::ProvablyFaultyBlock(_)) = proof_blocks.first() {
```
I don't understand this line. You got a block with a `MisbehaviorReport` and `MisbehaviorProof` inside it, which means that it was deserialized as `MisbehaviorProof::InvalidBlock(something)`. But where do you verify that `something` is actually a `BlockRef` of an invalid block?

cyberphysic4l replied:
The `proof_blocks` are retrieved from the local DAG state, so the `GetBlockResult::ProvablyFaultyBlock(_)` being returned by the dag state means that the block corresponding to this `BlockRef` in the dag state is saved as provably faulty. That means the block itself was received locally and failed post-signature verification, i.e. it is provably faulty.

cyberphysic4l replied:
I will elaborate on this in the comment to make it more clear if that now makes sense.

oliviasaa commented on consensus/core/src/block_verifier.rs:
```diff
@@ -237,6 +247,68 @@ impl BlockVerifier for SignedBlockVerifier {
+                    for proof_block in proof_blocks {
+                        if let GetBlockResult::VerifiedBlock(proof_block) = proof_block {
+                            if proof_block.epoch() != self.context.committee.epoch() {
+                                return Err(ConsensusError::InvalidEquivocationReport(index));
```
I think you can remove this line :)

oliviasaa commented on consensus/core/src/block_verifier.rs:
```diff
@@ -237,6 +247,68 @@ impl BlockVerifier for SignedBlockVerifier {
         Ok(())
     }
+
+    fn verify_misbehavior_reports(
```
I'm missing the `target` verification and the ordering between `first` and `second` in the equivocation part

oliviasaa commented:
A general question, because sometimes it is a bit difficult to follow the whole flow:
A block containing an invalid `MisbehaviorReport` is also provably faulty, right?

cyberphysic4l commented:
> A general question, because sometimes it is a bit difficult to follow the whole flow: A block containing an invalid `MisbehaviorReport` is also provably faulty, right?

Actually no, it is treated like a block with rejected ancestors, it will be rejected itself, but it is not provably faulty. The `verify()` function that determines if a block is provably faulty does not involve checking the ancestors (or references in misbehavior reports) because that requires retrieving those blocks.

cyberphysic4l commented:
Provable misbehavior reporting is deferred until we have a stable version of starfish 
See relevant issue https://github.com/iotaledger/iota-private/issues/180
